crates/qpy/src/circuits.rs

```diff
@@ -0,0 +1,2166 @@
+            if creg.name() == name {
+                register = Some(creg);
+            }
+        }
+        match register {
+            Some(register) => Ok(register.into_py_any(py)?),
+            None => Err(PyValueError::new_err(format!(
+                "Could not find classical register {:?}",
+                name
+            ))),
+        }
+    }
+}
+
+fn get_condition_data_from_inst(
+    py: Python,
+    inst: &Py<PyAny>,
+    circuit_data: &CircuitData,
+    qpy_data: &QPYWriteData,
+) -> PyResult<formats::ConditionPack> {
+    match getattr_or_none(inst.bind(py), "_condition")? {
+        None => Ok(formats::ConditionPack {
+            key: formats::condition_types::NONE,
+            register_size: 0u16,
+            value: 0i64,
+            data: formats::ConditionData::None,
+        }),
+        Some(condition) => {
+            if condition.extract::<classical::expr::Expr>().is_ok() {
+                let expression = pack_generic_data(&condition, qpy_data)?;
+                Ok(formats::ConditionPack {
+                    key: formats::condition_types::EXPRESSION,
+                    register_size: 0u16,
+                    value: 0i64,
+                    data: formats::ConditionData::Expression(expression),
+                })
+            } else if condition.is_instance_of::<PyTuple>() {
+                let key = formats::condition_types::TWO_TUPLE;
+                let value = condition
+                    .downcast::<PyTuple>()?
+                    .get_item(1)?
+                    .extract::<i64>()?;
+                let register =
+                    dumps_register(condition.downcast::<PyTuple>()?.get_item(0)?, circuit_data)?;
+                Ok(formats::ConditionPack {
+                    key,
+                    register_size: register.len() as u16,
+                    value,
+                    data: formats::ConditionData::Register(register),
+                })
+            } else {
+                Err(PyValueError::new_err(
+                    "Expression handling not implemented for get_condition_data_from_inst",
+                ))
+            }
+        }
+    }
+}
+
+fn get_condition_data(
+    py: Python,
+    op: &PackedOperation,
+    circuit_data: &CircuitData,
+    qpy_data: &QPYWriteData,
+) -> PyResult<formats::ConditionPack> {
+    match op.view() {
+        OperationRef::Instruction(py_inst) => {
+            get_condition_data_from_inst(py, &py_inst.instruction, circuit_data, qpy_data)
+        }
+        // we assume only PyInstructions have condition data at this stage
+        _ => Ok(formats::ConditionPack {
+            key: formats::condition_types::NONE,
+            register_size: 0u16,
+            value: 0i64,
+            data: formats::ConditionData::None,
+        }),
+    }
+}
+
+fn recognize_custom_operation(
+    py: Python,
+    op: &PackedOperation,
+    name: &String,
+) -> PyResult<Option<String>> {
+    let library = py.import("qiskit.circuit.library")?;
```

Comment:
I wonder if we can rely on `op.view()` here since custom gates would show up as one of the `Py*` variants anyway.